packages/block-library/src/template-part/edit/index.js

```diff
@@ -27,44 +27,51 @@ export default function TemplatePartEdit( {
 	setAttributes,
 	clientId,
 } ) {
-	const initialPostId = useRef( _postId );
 	const initialSlug = useRef( slug );
 	const initialTheme = useRef( theme );
 
 	// Resolve the post ID if not set, and load its post.
 	const postId = useTemplatePartPost( _postId, slug, theme );
 
-	// Set the post ID, once found, so that edits persist,
-	// but wait until the third inner blocks change,
-	// because the first 2 are just the template part
-	// content loading.
-	const { innerBlocks } = useSelect(
+	// Set the postId attribute if it did not exist,
+	// but wait until the inner blocks have loaded to allow
+	// new edits to trigger this.
+	const { innerBlocks, entityRecordEdits } = useSelect(
 		( select ) => {
 			const { getBlocks } = select( 'core/block-editor' );
+			const { getEntityRecordEdits } = select( 'core' );
 			return {
 				innerBlocks: getBlocks( clientId ),
+				entityRecordEdits: getEntityRecordEdits(
+					'postType',
+					'wp_template_part',
+					postId
+				),
 			};
 		},
-		[ clientId ]
+		[ clientId, postId ]
 	);
 	const { editEntityRecord } = useDispatch( 'core' );
-	const blockChanges = useRef( 0 );
+	const [ areInnerBlocksLoaded, setAreInnerBlocksLoaded ] = useState( false );
 	useEffect( () => {
-		if ( blockChanges.current < 4 ) blockChanges.current++;
+		if ( postId === null || postId === undefined ) {
+			return;
+		}
+
+		if ( ! areInnerBlocksLoaded ) {
+			if ( innerBlocks === entityRecordEdits?.blocks ) {
```

Comment:
We are considering innerBlocks to be loaded when they equal the blocks on file for `entityRecordEdits`.  This seems to work well in practice, but it is a bit of a coincidence that the innerBlocks updating triggers the recordEdits and that the values are not equal until the innerBlocks are loaded.

Alternatively, it would make more sense to compare innerBlocks to the raw content parsed from `getEntityRecord`.  However, this would require us parsing the raw content into blocks and doing a deep comparison on the arrays.  The approach with `entityRecordEdits` is a bit less intuitive, but it saves us from requiring extra parsing and evaluations.